### ISSUE TITLE
Problems when running PME over source which has already had PME ran over it

### ISSUE THREAD

**Issue by paulgallagher75** [release-engineering/pom-manipulation-ext#438]
If you run PME over the source twice there is a couple of things that may Error and 1 duplicate bit of info added to a pom file
1. If using the xmlUpdate -
   ERROR o.c.m.ext.manip.ManipulatingEventSpy - Extension failure
   org.commonjava.maven.ext.manip.ManipulationException: Did not locate XML using XPath //include[starts-with(.,'org.apache.tomcat')]
2. If using the jsonUpdate - 
   ERROR o.c.m.ext.manip.impl.JSONManipulator - XPath $..resolved did not find any expressions within artemis-website/npm-shrinkwrap.json 
   ERROR o.c.m.ext.manip.ManipulatingEventSpy - Extension failure
   org.commonjava.maven.ext.manip.ManipulationException: XPath did not resolve to a valid value
3. the top level pom file has an extra comment added, this is identical to the one already there -
   
   ```
   <!--
   Modified by POM Manipulation Extension for Maven 2.5 ( SHA: 9802ee66 ) 
   -->
   ```

**Comment by jdcasey** [comment]
So, would a better solution be to output a warning if the (now replaced?) path isn't found for #'s 1 & 2?

For #3, I suppose we would need to track down the original comment PME added and either leave it in place (...which, if Maven version has changed or something,might be wrong), or replace it.

**Comment by rnc** [comment]
1/2 ) For both of these, on the original transformation we probably do want an error if we don't find anything.  Editing the configuration to not include jsonUpdate/xmlUpdate for the second pass is a solution (though awkward). Automatically detecting the PME has run before and we don't need to emit an error is relatively tricky as this is being run on a clean checkout (so there are no marker files). In fact, the only marker is from (3). Not sure if that can be used to change error->warning (plus might be fairly invasive).

3) That sounds like a bug. PomIO should not add duplicate comment lines.

**Comment by rnc** [comment]
@paulgallagher75  I can reproduce number 3. For 1/2 would examining if PME has been run before on the sources work for your use-case?

**Comment by rnc** [comment]
@paulgallagher75 @jdcasey For (1/2) ..... by using the method suggested above i.e. examine the root execution pom, we can emit an exception on first pass. For subsequent passes - do you have any opinions on whether this should be configurable or not? Currently I have it non-configurable - subsequent runs will not throw an error but just emit a warning.

**Comment by jdcasey** [comment]
@rnc having it configurable definitely has precedent. For instance, strict checksum checking, fail vs. warn modes in surefire, failsafe, and enforcer plugins. I'm not sure how hard that would be, or if we have both use cases right now...but you could argue the principle of least surprise in favor of allowing configuration.

Having said that, I'd say it's whatever matches our existing use cases. There's nothing to prevent re-adding the failure option later.

**Comment by rnc** [comment]
@jdcasey I think my preference would be to keep the prior behaviour where JSON/XML manipulators throws an exception on a pristine repository.  I think that in this particular scenario, a warning for subsequent runs make sense. If the original JSON/XML request was to e.g. remove a line or update a line that expression may not be found in subsequent runs.  As you say, if this does prove problematic we can always make it configurable to choose to warn or throw an exception on subsequent runs in future.

**Comment by rnc** [comment]
I am assigning this to @paulgallagher75 to await testing.

**Comment by jdcasey** [comment]
+1

**Comment by rnc** [comment]
Fixed by #440
